Markdown parser

shamshirz opened the pull request:
Gimme some thoughts, it works, but before I create the tests (i've been testing straight in the browser, sorry bad practice) and the typespec I'd love some thoughts.

shamshirz committed:
Add parsing functions, no typespec yet

shamshirz committed:
Add template for showing parsed data

rjdellecese commented on web/templates/quote/show.html.eex:
```diff
@@ -5,6 +5,28 @@
+    <strong>Parsed</strong>
+    <table class="table">
+      <tbody>
+      <%= for style <- parse(@quote.raw) do %>
```
Nice comprehension!

rjdellecese commented on web/views/quote_view.ex:
```diff
@@ -1,3 +1,68 @@
+  # This captures(named style) on a caret followed by EITHER one of the characters, or 3 digits
+  @markdown_regex ~r/\^((?<style>(\d{3})|[wrbymnpkotvg*]))/i
+
+  def parse(quote) do
```
What about something like this?
```Elixir
def parse(quote) do
  Regex.split(~r/\^[^\^]*/, quote.raw, trim: true, include_captures: true)
  |> Enum.map(&create_style_struct/1)
end

def create_style_struct(substring) do
  cond do
    String.at(substring, 0) != "^" ->
      %Style{color: nil, body: substring}
    String.slice(substring, 1..3) =~ ~r/\d{3}/ ->
      %Style{color: get_color_by_rgb(String.slice(substring, 1..3)),
             body: String.slice(substring, 3..-1)}
    String.at(substring, 1) =~ ~r/[wrbymnpkotvg*]/i ->
      %Style{color: get_color_by_code(String.at(substring, 1)),
             body: String.slice(substring, 2..-1)}
    true ->
      %Style{color: nil, body: substring}
  end
end
```
Also if we use a struct we'd also need to define it in it's own module, I think, e.g.
```Elixir
defmodule Style
  @enforce_keys [:body]
  defstruct [:color, :body]
end
```

shamshirz replied:
interesting, not opposed. I'll look into doing both of these next

rjdellecese commented on web/views/quote_view.ex:
```diff
@@ -1,3 +1,68 @@
+      "t" -> "teal"
+      "v" -> "grey"
+      "g" -> "green"
+      _   -> "didn't match anything"
```
I think this case should probably not always match... I think it would probably be better to handle no-matches on color codes upstream, and assume that if no match occurs here there was an error.

rjdellecese commented on web/views/quote_view.ex:
```diff
@@ -1,3 +1,68 @@
+    color = parse_markdown(string)
+    body = get_without_markdown(string) |> String.trim
+    IO.puts "style is #{color} and body is #{body}"
+    %{color: color, body: body}
```
I think we should ask Sam's opinion in a little more detail about how `:color` should be formatted.

rjdellecese reviewed (commented):
Dopee Elixir is dopee!

In my API branch I just got rid of the browser endpoints, because it was going to be more work to keep them around than it was going to be to get rid of them (basically because I wanted to use that same `QuoteController` and not have to handle both request types through it), so I'd say don't bother doing any more HTML templating.

Lemme know what you think about my comments!

rjdellecese commented on web/views/quote_view.ex:
```diff
@@ -1,3 +1,68 @@
+  # Splits at every caret, and includes the caret.
+  # returns a list of strings with a single markdown style in them
+  def split_on_markdown(string) do
+    string_styles = Regex.split(~r/\^.*?(?=\^)/ , string, trim: true, include_captures: true)
```
Ahaha, in the comment above I think `Regex.split(~r/\^[^\^]*/, quote.raw, trim: true, include_captures: true)` is functionally (nice punned!) equivalent to this line...

rjdellecese replied:
So, the regex is.

shamshirz replied:
word, what do the brackets mean?

rjdellecese replied:
Brackets mean "this set of characters" and so since the first caret in that bracket means "not the following character, the regex is saying "match a caret and any number of any character except another caret".

rjdellecese commented on web/views/quote_view.ex:
```diff
@@ -1,3 +1,68 @@
   use HonGraffitiPhoenix.Web, :view
+
+  # This captures(named style) on a caret followed by EITHER one of the characters, or 3 digits
+  @markdown_regex ~r/\^((?<style>(\d{3})|[wrbymnpkotvg*]))/i
```
Markdown is technically a specific type of markup, so I think we should use either `markup` or maybe even `hon_markup` instead.

shamshirz replied:
yes, will do that, should have read these comments before making todays changes.

shamshirz committed:
Move markdown_parser into lib, update views and templates

shamshirz committed:
Add tests for the markdown_parser

shamshirz committed:
Update var names, color_code parser, regex in parser

shamshirz commented:
linting is next and adding the typespec stuff

shamshirz committed:
Add parser and tests

Added a new directory under lib for parsing tools, and
a struct for use with the parser

shamshirz committed:
Fix web parser alias

shamshirz committed:
Simplify parser and tests

shamshirz committed:
Added parsing color by rgb value

shamshirz committed:
Update module name to markup from markdown

shamshirz committed:
Update another markdown to markup

shamshirz commented:
Hey, I've got some changes @rjdellecese 
1. Let me know what you think about the overall structure
2. I didn't throw on not matching a color, at what point should we?

shamshirz committed:
Update function name in test

rjdellecese commented on lib/hon_graffiti_phoenix/parsers/markup_parser.ex:
```diff
@@ -0,0 +1,64 @@
+defmodule HonGraffitiPhoenix.Parsers.MarkupParser do
+  alias HonGraffitiPhoenix.Parsers.DecoratedString
+
+  @color_map %{
```
Nice!

rjdellecese commented on lib/hon_graffiti_phoenix/parsers/markup_parser.ex:
```diff
@@ -0,0 +1,64 @@
+  def parse_markup_code(markup_code) do
+    case String.length(markup_code) do
+      x when x == 1 -> Map.get(@color_map, markup_code, "black")
+      x when x > 1  -> Enum.join(["rgb(", get_color_by_rgb(markup_code), ")"])
```
I just realized this `case` uses those built-in guard clauses, pretty cool.

rjdellecese commented on lib/hon_graffiti_phoenix/parsers/markup_parser.ex:
```diff
@@ -0,0 +1,64 @@
+
+  defp get_color_by_rgb(rgb_string) do
+    rgb_string
+    |> String.graphemes
```
Not really a comment on your code, but I like how technically precise this function name is, as opposed to something like `chars` (which would be the Ruby equivalent, at least).

shamshirz replied:
agreed, super excited to find this

rjdellecese commented on lib/hon_graffiti_phoenix/parsers/markup_parser.ex:
```diff
@@ -0,0 +1,64 @@
+  def parse(quote) do
+    quote
+    |> String.split(@caret_regex, trim: true, include_captures: true)
+    |> Enum.map(fn segment -> parse_segment(segment) end)
```
I think
```Elixir
|> Enum.map(&parse_segment/1)
```
is the syntax you were looking for yesterday!

shamshirz replied:
nailed it, thanks!

rjdellecese commented on lib/hon_graffiti_phoenix/parsers/markup_parser.ex:
```diff
@@ -0,0 +1,64 @@
+    rgb_string
+    |> String.graphemes
+    |> Enum.map(&String.to_integer(&1) * 28)
+    |> Enum.join(",")
```
What do you think about putting this line (which is currently in `parse_markup_code/1`)
```Elixir
x when x > 1  -> Enum.join(["rgb(", get_color_by_rgb(markup_code), ")"])
```
in this function? Maybe something like
```Elixir
  defp get_color_by_rgb(rgb_string) do
    comma_separated_rgb_values = rgb_string
    |> String.graphemes
    |> Enum.map(&String.to_integer(&1) * 28)
    |> Enum.join(",")
    Enum.join(["rgb(", comma_separated_rgb_values, ")"])
  end
```

rjdellecese replied:
Also, you could write that last line as `"rgb(#{comma_separated_rgb_values})"` (or even `"rgb(" <> comma_separated_rgb_values <> ")"`). In Ruby the string interpolation approach would be considered better style (also better style than `join`ing, I think), but I'm not sure which approach would be preferred in Elixir.

Mostly just thinking out loud/wondering what the most "correct" way would be...

shamshirz replied:
Okay, I like how get_color_by_rgb is just the pipe chain, I think that makes it clean and readable. I replaced the Enum.join in the markup_parser to the string interpolation. Cheggit outi (push incoming)

rjdellecese replied:
IMO it's more valuable to have the boundaries between the functions more clearly defined conceptually. Plus, the flow of data is still pretty straightforward I think, even though it happens to not use all pipes in that case. But up to you :)

shamshirz replied:
BOOM
```Elixir
defp get_color_by_rgb(rgb_string) do
    rgb_string
    |> String.graphemes
    |> Enum.map(&String.to_integer(&1) * 28)
    |> Enum.join(",")
    |> (&("rgb(" <> &1 <> ")")).()
  end
```

rjdellecese replied:
I did some research on that sort of syntax last night, the Elixir community seems to have come to some consensus that that's bad style: https://elixirforum.com/t/pipe-operator-and-functions-not-function-calls/782

Personally I agree, but I won't complain if Credo lets it through :p

rjdellecese replied:
Also this discussion in the mailing list: https://groups.google.com/forum/#!topic/elixir-lang-core/WDePSnaiR5w

shamshirz replied:
Okay, made a named function 1 liner that wraps it. It's private and only used once, seems like the elixer-y way of doing it

rjdellecese replied:
Love it

rjdellecese commented on lib/hon_graffiti_phoenix/parsers/decorated_string.ex:
```diff
@@ -0,0 +1,8 @@
+  """
+
+  @enforce_keys [:body]
+  defstruct [:body, color: "black"]
```
Why default to black?

shamshirz replied:
Because I was thinking about a normal white background, should be white?

rjdellecese replied:
Aye, HoN default is white hashtagpmlife, so the background will be some kind of dark or opaque black (like the chat box in-game).

rjdellecese commented on lib/hon_graffiti_phoenix/parsers/markup_parser.ex:
```diff
@@ -0,0 +1,64 @@
+    "b" => "blue", "y" => "yellow", "m" => "magenta",
+    "n" => "brown", "p" => "purple", "o" => "orange",
+    "t" => "teal", "v" => "grey", "g" => "green",
+    "k" => "black", "" => "black"}
```
Re: this empty string key, I think if a caret is followed by an invalid color code this map should never be accessed and the caret should be rendered in the body (so this key/value pair should be unnecessary).

As a separate but related point, it looks like you have a test that verifies the inverse of that ("broken markdown won't be parsed into a color") but I think it might also be good to verify that the body includes the caret, as expected.

rjdellecese replied:
Actually, I take that back, we need to test this in HoN, I'm not sure that that's the behavior that occurs... Either way, though, per your question about when to raise, I don't think we should as I don't think HoN would.

shamshirz replied:
swaq, not going to deal with it now

rjdellecese replied:
Cool, I'll look into this when I get the chance.

rjdellecese commented on lib/hon_graffiti_phoenix/parsers/markup_parser.ex:
```diff
@@ -0,0 +1,64 @@
+  If no markup is found, a default is given.
+  """
+  @spec parse_segment(String.t) :: meta_string
+  def parse_segment(string) do
```
Can this and `parse_markup_code` be private?

shamshirz replied:
Yes they can! The reason I didn't tho, was because I wanted to unit test them...Those Hats Over Theres?

shamshirz replied:
Nevermind! I just made all of the tests use the upper functions, private away!

rjdellecese replied:
Awesome, IMO that's the ideal 👍  Get dem interfaces expressing intent, nam sayin?

rjdellecese reviewed (commented):
Sorry if I left a bunch of comments on things you were going to address, if so just ignore them. Overall I think this is looking great, and is easier to read with the recent changes you made to the whole flow of data!

shamshirz committed:
Updated functions and tests from PR feedback

Added some tests, defaults color to white
streamlined functions, made them more elixiry

shamshirz committed:
Update get_color_by_rgb to return entire rgb string

shamshirz committed:
Replace anon fxn with named fxn

rjdellecese commented on test/lib/markdown_parser_test.exs:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  test "it can parse 3 digit codes or single characters" do
+    assert hd(MarkupParser.parse("^000")).color == "rgb(0,0,0)"
```
Cool `hd` Kernel function bro.

shamshirz replied:
sarcasm? Is there a way to use the [h | t] notation to shorten?

rjdellecese replied:
No /s, it's cool!

rjdellecese commented on lib/hon_graffiti_phoenix/parsers/decorated_string.ex:
```diff
@@ -0,0 +1,8 @@
+defmodule HonGraffitiPhoenix.Parsers.DecoratedString do
+  @moduledoc """
+  Provides a struct for strings with hon style markdown
```
Gonna make literally the most annoying request: can you change this to `markup`?

shamshirz replied:
just got did

rjdellecese reviewed (changes requested):
See "literally most annoying request" and then I will merge! Thanks for all ur hustle on this homie!

shamshirz committed:
Replace doc comment about markdown w/ markup

shamshirz commented:
omg omg omg this actually gunna get Pult?

shamshirz commented:
HOLY SHITTTEEEE
